### PR TITLE
8340547: Starting many threads can delay safepoints

### PR DESCRIPTION
**Pull request by olivergillespie**
Mitigate the impact of JVM_StartThread on safepoint synchronization, by adding a new ThreadStart_lock which limits the number of JVM_StartThread invocations competing for Threads_lock at any given time to 1.
This gives a VM thread trying to call a safepoint a much better chance of acquiring Threads_lock when there are many JVM_StartThread invocations in flight, at the cost of one extra lock/unlock for every new thread.

Can be disabled with new diagnostic flag `-XX:-UseExtraThreadStartLock`.

Before (ThreadStartTtsp.java is shared in JDK-8340547):
```
java -Xlog:safepoint ThreadStartTtsp.java | grep -o 'Reaching safepoint: [0-9]* ns'
Reaching safepoint: 1291591 ns
Reaching safepoint: 59962 ns
Reaching safepoint: 1958065 ns
Reaching safepoint: 14456666258 ns <-- 14 seconds!
...
```

After:
```
java -Xlog:safepoint ThreadStartTtsp.java | grep -o 'Reaching safepoint: [0-9]* ns'
Reaching safepoint: 214269 ns
Reaching safepoint: 60253 ns
Reaching safepoint: 2040680 ns
Reaching safepoint: 3089284 ns
Reaching safepoint: 2998303 ns
Reaching safepoint: 4433713 ns <-- 4.4ms
Reaching safepoint: 3368436 ns
Reaching safepoint: 2986519 ns
Reaching safepoint: 3269102 ns
...
```

**Alternatives**

I considered some other options for mitigating this. For example, could we reduce the time spent holding the lock in StartThread? Most of the time is spent managing the threads list for ThreadSMR support, and each time we add a thread to that list we need to copy the whole list and free every entry in the original, which is slow. But I didn't see an easy way to avoid this.
I also looked at some kind of signal from the VM thread that it is ready to start synchronizing that StartThread could check before trying to grab Threads_lock, but I didn't find anything better than this extra lock.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340547](https://bugs.openjdk.org/browse/JDK-8340547): Starting many threads can delay safepoints (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) Review applies to [7b5633ae](https://git.openjdk.org/jdk/pull/21111/files/7b5633ae86f09710123458d2eefa1912394ced16)
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21111/head:pull/21111` \
`$ git checkout pull/21111`

Update a local copy of the PR: \
`$ git checkout pull/21111` \
`$ git pull https://git.openjdk.org/jdk.git pull/21111/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21111`

View PR using the GUI difftool: \
`$ git pr show -t 21111`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21111.diff">https://git.openjdk.org/jdk/pull/21111.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21111#issuecomment-2364005855)

### CODE REVIEW

File: src/hotspot/share/runtime/globals.hpp [hunk 1]
```diff
@@ -1997,6 +1997,10 @@ const int ObjectAlignmentInBytes = 8;
                                                                             \
   product(bool, StressSecondarySupers, false, DIAGNOSTIC,                   \
           "Use a terrible hash function in order to generate many collisions.") \
+                                                                            \
+  product(bool, UseThreadStartLock, true, DIAGNOSTIC,                       \
+          "Use an extra lock in JVM_ThreadStart to reduce "                 \
+          "time-to-safepoint while creating threads.")                      \
```
Comment (shipilev) [inline src/hotspot/share/runtime/globals.hpp:2003]: I think the flag can be just `UseThreadStartLock`. The description should probably be more generic, e.g.:

```
Use an extra lock during Thread.start to alleviate contention on Threads lock.
```

**Comment by dholmes-ora** [comment]
What is the performance hit in your scenario of starting many threads?

**Comment by olivergillespie** [comment]
> What is the performance hit in your scenario of starting many threads?

Do you mean what is the overhead of creating the threads themselves, compared to the time-to-safepoint issue? I'm not sure how best to measure that. It's expensive (especially with SMR), and obviously an antipattern, but you would usually expect the cost of starting threads to be mostly localized to the caller compared to this issue which affects the whole vm in an unexpected way.

**Comment by dholmes-ora** [comment]
> Do you mean what is the overhead of creating the threads themselves, compared to the time-to-safepoint issue?

Yes. By adding the additional layer of throttling you've addressed your TTSP issue but at the expense of slowing down the overall rate of thread creation. So what is the likely overhead for applications that create lots of threads (at a relatively fast rate) but which don't see a TTSP issue?

**Comment by olivergillespie** [comment]
Oh I see what you mean now, sorry - what penalty does the fix add to the normal case? On my x86 host:

```
@Benchmark
public Thread startThread() {
    Thread t = new Thread(() -> {});
    t.start();
    return t;
}
```

With fix: 16285 ± 189 op/s.
Without fix: 16375 ± 158  ops/s

So possibly within the noise, or perhaps somewhere around ~0.5% overhead. It's the cost of acquiring one extra lock, which intuitively is negligible compared to the significant cost of starting a thread with SMR. Starting threads is already slow, adding a little to that to avoid hard to debug vm-wide delays seems sensible to me.

**Comment by shipilev** [comment]
I used to worry about scalability impact of this change, i.e. contending on a new lock, but quickly realized we are likely contending on `ThreadLock` in that scenario. I suspect this just shifts the contention from `ThreadLock` to `StartThreadLock` when multiple threads are starting up, so it makes sense if we do not see much of the impact on `Thread.start`.

You need to also test perf with multiple threads doing `Thread.start`, @olivergillespie -- so that these locks be contended. Your reproducer already has the kernel for it, maybe just measure the thread starting speed (in threads/sec) before and after this patch?

**Comment by olivergillespie** [comment]
With a slightly modified version of my reproducer (removing the GC triggers):

```
import java.util.concurrent.Semaphore;
import java.util.concurrent.ThreadLocalRandom;

public class ThreadStartTtsp {
    static int NUM_THREADS = 1_000;

    public static void main(String[] args) throws InterruptedException {
        Semaphore s = new Semaphore(NUM_THREADS);

        // Start lots of threads at the same time to cause contention on freelist
        for (int i = 0; i < NUM_THREADS; i++) {
            new Thread(() -> {
                while (true) {
                    try { s.acquire(); } catch (Exception e) { throw new RuntimeException(e); }
                    new Thread().start();
                }
            }).start();
        }

        // Periodically 'trigger' the thread creation
        for (int i = 0; i < 100; i++) {
            s.release(NUM_THREADS);
            long start = System.nanoTime();
            while (s.availablePermits() != 0) { }
            System.out.printf("Started %d threads in %.3fms%n", NUM_THREADS, (System.nanoTime() - start)/1_000_000.0);
        }
    }
}
 ```
 
 I see higher throughput with the new lock enabled compared to disabled.
 
 ```
 java -XX:+UnlockDiagnosticVMOptions -XX:-UseThreadStartLock ThreadStartTtsp.java
...
Started 1000 threads in 134.874ms
Started 1000 threads in 134.799ms
Started 1000 threads in 132.662ms
...
```

```
./build/release/images/jdk/bin/java -XX:+UnlockDiagnosticVMOptions -XX:+UseThreadStartLock ThreadStartTtsp.java
...
Started 1000 threads in 89.469ms
Started 1000 threads in 90.030ms
Started 1000 threads in 89.855ms
...
```

I think that's because of reduced contention with `JavaThread::exit`, not sure if it's a fair comparison.

Probably fairer is this:

```
import java.util.concurrent.atomic.*;

public class ThreadStartTtsp {
    static int NUM_THREADS = 1_00;
    static AtomicInteger started = new AtomicInteger(0);
    static long start = System.nanoTime();

    public static void main(String[] args) throws InterruptedException {
        for (int i = 0; i < NUM_THREADS; i++) {
            new Thread(() -> {
                while (true) {
                    new Thread(() -> {
                        if (started.incrementAndGet() % 10_000 == 0) {
                            System.out.printf("Started %d threads in %.3fms%n", 10_000, (System.nanoTime() - start)/1_000_000.0);
                            start = System.nanoTime();
                        }
                    }).start();
                }
            }).start();
        }
    }
}
```

For which I still see about 4% improvement from adding the new lock, and the gap seems to widen as concurrency increases.

**Comment by merykitty** [comment]
> I considered some other options for mitigating this. For example, could we reduce the time spent holding the lock in StartThread? Most of the time is spent managing the threads list for ThreadSMR support, and each time we add a thread to that list we need to copy the whole list and free every entry in the original, which is slow. But I didn't see an easy way to avoid this.
I also looked at some kind of signal from the VM thread that it is ready to start synchronizing that StartThread could check before trying to grab Threads_lock, but I didn't find anything better than this extra lock.

Maybe you have already considered it but an alternative is to use a priority lock.

**Comment by olivergillespie** [comment]
> Maybe you have already considered it but an alternative is to use a priority lock.

Thanks. Is there an implementation available in hotspot?

**Comment by dean-long** [comment]
If JVM_StartThread is only called by Thread.start0, then how about putting the new lock in Java instead?

**Comment by olivergillespie** [comment]
> If JVM_StartThread is only called by Thread.start0, then how about putting the new lock in Java instead?

What benefit do you see of that? One downside is that the lock will be coarser than necessary. I'd rather keep the lock as tightly scoped as possible.

**Comment by olivergillespie** [comment]
The same kind of issue happens on thread exit, via `JavaThread::post_run -> JavaThread::exit -> Threads::remove ->  Threads_lock.lock`, so I will update the implementation to cover both with the same approach.

**Comment by dholmes-ora** [comment]
If you do this at exit too then my concerns have doubled. I'd want to see some broader benchmarking of this change.

**Comment by olivergillespie** [comment]
My targeted benchmarks show either significant improvements or negligible slowdown on a (typically, and arguably always if performance is key) non-critical path, which matches the theory since we're not adding any contention, only adding one extra monitor acquisition on a slow path.
Hopefully that's enough to indicate reasonable performance, so we can finalize the implementation and then perhaps do some final benchmarking before integration?

**Comment by olivergillespie** [comment]
Added `ThreadStopTtsp.java` reproducer to JDK-8340547 for the thread exit case, and updated the implementation here to cover both cases.

File: src/hotspot/share/runtime/globals.hpp [hunk 2]
```diff
@@ -1997,6 +1997,10 @@ const int ObjectAlignmentInBytes = 8;
                                                                             \
   product(bool, StressSecondarySupers, false, DIAGNOSTIC,                   \
           "Use a terrible hash function in order to generate many collisions.") \
+                                                                            \
+  product(bool, UseThreadsLockThrottleLock, true, DIAGNOSTIC,               \
+          "Use an extra lock during Thread start and exit to alleviate"     \
+          "contention on threads lock.")                                    \
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/globals.hpp:2003]:
```suggestion
          "contention on Threads_lock.")                                    \
```

File: src/hotspot/share/runtime/mutexLocker.hpp [hunk 3]
```diff
@@ -62,6 +62,8 @@ extern Monitor* CodeCache_lock;                  // a lock on the CodeCache
 extern Mutex*   TouchedMethodLog_lock;           // a lock on allocation of LogExecutedMethods info
 extern Mutex*   RetData_lock;                    // a lock on installation of RetData inside method data
 extern Monitor* VMOperation_lock;                // a lock on queue of vm_operations waiting to execute
+extern Monitor* ThreadsLockThrottle_lock;        // used by Thread start/stop to reduce competition for Threads_lock,
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/mutexLocker.hpp:65]:
```suggestion
extern Monitor* ThreadsLockThrottle_lock;        // used by Thread start/exit to reduce competition for Threads_lock,
```

**Review by dholmes-ora — COMMENTED** [review]
Implementation looks good. Lets see what the benchmarking shows. We can reason that for applications that create lots of threads quickly, this additional throttling can actually improve general throughput. But of course any individual start/exit is slowed down by the extra lock acquisition and release.

Thanks

**Comment by olivergillespie** [comment]
Thanks @dholmes-ora . Is there any particular benchmarking you recommend me to run?

**Comment by dean-long** [comment]
> > If JVM_StartThread is only called by Thread.start0, then how about putting the new lock in Java instead?
> 
> What benefit do you see of that? One downside is that the lock will be coarser than necessary. I'd rather keep the lock as tightly scoped as possible.

I just thought it would be simpler, but I see your point.  A coarser lock will serialize more of the native path.

**Review by dholmes-ora — APPROVED** [review]
Looks fine. Just confirm general benchmarking results. Thanks

**Review by shipilev — APPROVED** [review]
Looks fine, only cosmetics:

**Comment by olivergillespie** [comment]
> Looks fine. Just confirm general benchmarking results. Thanks

Thank you. No observed regressions in my testing, including SpecJBB, SpecJVM and DaCapo.

**Review by shipilev — APPROVED** [review]
Looks fine to me, thanks.

**Comment by shipilev** [comment]
I think @dean-long and @merykitty might need to approve or indicate they do not have other comments without the formal review.

**Review by merykitty — APPROVED** [review]
I think the patch looks good.

**Comment by shipilev** [comment]
@dean-long -- do you have any more comments about this?

**Comment by shipilev** [comment]
All right, if Dean has more comments, we can handle them after the integration.

/sponsor

**Pull request closed without merging**